barretenberg/cpp/src/barretenberg/vm/avm/generated/circuit_builder.cpp

```diff
@@ -30,15 +49,48 @@ AvmCircuitBuilder::ProverPolynomials AvmCircuitBuilder::compute_polynomials() co
                                               /*make shiftable with offset*/ 1 };
                        }
                    }));
+
     // catch-all with fully formed polynomials
     AVM_TRACK_TIME(
         "circuit_builder/init_polys_unshifted", ({
             auto unshifted = polys.get_unshifted();
-            bb::parallel_for(unshifted.size(), [&](size_t i) {
+
+            // An array which stores for each column of the trace the smallest size of the
+            // truncated column containing all non-zero elements.
+            // It is used to allocate the polynomials without memory overhead for the tail of zeros.
+            std::array<size_t, Row::SIZE> col_nonzero_size{};
+
+            // Computation of size of columns.
+            // Non-parallel version takes 0.5 second for a trace size of 200k rows.
+            // A parallel version might be considered in the future.
+            for (size_t i = 0; i < num_rows; i++) {
+                const auto row = rows[i].as_vector();
+                for (size_t j = 0; j < Row::SIZE; j++) {
+                    if (!row[j].is_zero()) {
+                        col_nonzero_size[j] = i + 1;
+                    }
+                }
+            }
+
+            // Set of the labels for derived/inverse polynomials.
+            const auto derived_labels = polys.get_derived_labels();
+            std::set<std::string> derived_labels_set(derived_labels.begin(), derived_labels.end());
+
+            bb::parallel_for(num_unshifted, [&](size_t i) {
                 auto& poly = unshifted[i];
+                const auto col_idx = polys_to_cols_unshifted_idx[i];
+                size_t col_size = 0;
+
+                // We fully allocate the inverse polynomials. We leave this potential memory optimization for later.
+                if (derived_labels_set.contains(labels[i])) {
```

Comment:
An idea (please double check): if you initialize `col_nonzero_size` to have `num_rows` on each slot, then you don't need to check the labels (and therefore also not construct them).

You do still need the poly_to_cols, unfortunatelly.